src/ring_theory/free_comm_ring.lean

```diff
@@ -0,0 +1,267 @@
+    { intros s ih, iterate 3 { rw free_abelian_group.lift.neg }, rw [ih, neg_add], refl },
+    { intros z1 z2 ih1 ih2, iterate 3 { rw free_abelian_group.lift.add }, rw [ih1, ih2],
+      rw [add_assoc, add_assoc], congr' 1, apply add_left_comm }
+  end,
+  .. free_comm_ring.add_comm_group α,
+  .. free_comm_ring.semigroup α }
+
+instance : comm_ring (free_comm_ring α) :=
+{ mul_comm := λ x y, begin
+    refine free_abelian_group.induction_on x (zero_mul y) _ _ _,
+    { intros s, refine free_abelian_group.induction_on y (zero_mul _).symm _ _ _,
+      { intros t, unfold has_mul.mul semigroup.mul ring.mul,
+        iterate 4 { rw free_abelian_group.lift.of }, rw add_comm },
+      { intros t ih, rw [mul_neg_eq_neg_mul_symm, ih, neg_mul_eq_neg_mul] },
+      { intros y1 y2 ih1 ih2, rw [mul_add, add_mul, ih1, ih2] } },
+    { intros s ih, rw [neg_mul_eq_neg_mul_symm, ih, neg_mul_eq_mul_neg] },
+    { intros x1 x2 ih1 ih2, rw [add_mul, mul_add, ih1, ih2] }
+  end
+  .. free_comm_ring.ring α }
+
+variables {α}
+def of (x : α) : free_comm_ring α :=
+free_abelian_group.of [x]
+
+section lift
+
+variables {β : Type v} [comm_ring β] (f : α → β)
+
+def lift : free_comm_ring α → β :=
+free_abelian_group.lift $ λ s, (s.map f).prod
+
+@[simp] lemma lift_zero : lift f 0 = 0 := rfl
+
+@[simp] lemma lift_one : lift f 1 = 1 :=
+free_abelian_group.lift.of _ _
+
+@[simp] lemma lift_of (x : α) : lift f (of x) = f x :=
+(free_abelian_group.lift.of _ _).trans $ mul_one _
+
+@[simp] lemma lift_add (x y) : lift f (x + y) = lift f x + lift f y :=
+free_abelian_group.lift.add _ _ _
+
+@[simp] lemma lift_neg (x) : lift f (-x) = -lift f x :=
+free_abelian_group.lift.neg _ _
+
+@[simp] lemma lift_sub (x y) : lift f (x - y) = lift f x - lift f y :=
+free_abelian_group.lift.sub _ _ _
+
+@[simp] lemma lift_mul (x y) : lift f (x * y) = lift f x * lift f y :=
+begin
+  refine free_abelian_group.induction_on y (mul_zero _).symm _ _ _,
+  { intros s2, conv { to_lhs, dsimp only [(*), mul_zero_class.mul, semiring.mul, ring.mul, semigroup.mul] },
+    rw [free_abelian_group.lift.of, lift, free_abelian_group.lift.of],
+    refine free_abelian_group.induction_on x (zero_mul _).symm _ _ _,
+    { intros s1, iterate 3 { rw free_abelian_group.lift.of }, rw [multiset.map_add, multiset.prod_add] },
+    { intros s1 ih, iterate 3 { rw free_abelian_group.lift.neg }, rw [ih, neg_mul_eq_neg_mul] },
+    { intros x1 x2 ih1 ih2, iterate 3 { rw free_abelian_group.lift.add }, rw [ih1, ih2, add_mul] } },
+  { intros s2 ih, rw [mul_neg_eq_neg_mul_symm, lift_neg, lift_neg, mul_neg_eq_neg_mul_symm, ih] },
+  { intros y1 y2 ih1 ih2, rw [mul_add, lift_add, lift_add, mul_add, ih1, ih2] },
+end
+
+instance : is_ring_hom (lift f) :=
+{ map_one := lift_one f,
+  map_mul := lift_mul f,
+  map_add := lift_add f }
+
+@[simp] lemma lift_pow (x) (n : ℕ) : lift f (x ^ n) = lift f x ^ n :=
+is_semiring_hom.map_pow _ x n
+
+theorem lift_unique (f : free_comm_ring α → β) [is_ring_hom f] : f = lift (f ∘ of) :=
```

Comment:
Wouldn't you state this equality the other way round? Maybe it is then a `simp`?

Reply:
done